_posts/2021-06-21-a-new-tale.md

```diff
@@ -0,0 +1,43 @@
+Additionally, the distributed nature of the new task worker type drops the use of `Redis Queue` (`rq`).
+With this simplification there is no longer a need to replicate task data accross two databases.
+The main Postgres Pulp database will therefore be the single source of truth about tasks.
+Last but not least, the new worker type will use Postgres advisory locks to ensure exclusive access to tasks.
+This is a much more lightweigt approach than storing access locks in a separate table.
+Also it simplifies the cleanup process vastly if a worker dies as postgres will clean the locks with its connection.
+
+With all these simplifications, the new tasking system is more reliable.
+Without the resource-manager as the single point of failure, it will be fully high available.
+Also with the latter seen as the choke point in task processing, we see, that the new distributed system will scale much better in performance with the number of parallel running workers.
+
+> ![Tasking system benchmark](/images/tasking_system_benchmark/scale_workers.png)
+  Benchmark of the old and new tasking system with a single thread bursting 256 tasks that need random 2 out of a pool of 128 abstract resources and need 0.5 s to complete.
+  Plotted is in the first panel the average time the foreground thread needs per task, and in the second the average time, a task needs from being available in the system and being picked up by a worker.
+  It can be seen, that in the old system, the `resource-manager` is able to deliver about 2 tasks per second, and that providing more than two workers do not provide additional speedup.
+  In the new tasking system however the tasks throughput scales well with the number of  workers provided.
+  However starting with 32 workers, the database load of concurrent processes polling the tasks table will slow down the insertion of new tasks into the system by a factor of about 5.
+
+The removal of the queueing system brings another nontrivial benefit to the table.
+As no tasks are assigned to a worker anymore before actually starting to be performed, there will be no tasks canceled that were already pending in the corresponding queue on a worker restart.
+
+This new system will be introduced in the upcoming 3.14 release of Pulpcore as a drop in replacement.
+With the current plan, it will be switched off by default.
+To activate the new system, you need to set the `USE_NEW_WORKER_TYPE` setting to `True`, and restart your Pulp services.
+For production environments, we recommend however, that you first drain your taskqueues, by shutting down the api-workers first.
+If you decide to go back, it is basically the same process, where you flip the setting to `False`.
+While this process introduces a downtime on the api side, the content delivery side is not affected by this change.
+The current plan is to mark the new system as default one minor Pulpcore release later, and deprecate the old system, to be removed yet another minor release after that.
```

Comment:
do we have which version this begins with?

Reply:
It's "one after" the aforementioned 3.14.